keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,261 @@
+
+## Table of Contents
+
+* [Terms](#terms)
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [User Stories](#user-stories)
+    * [Story 1 - NodeAffinity](#story-1---nodeaffinity)
+    * [Story 2 - PodAffinity](#story-2---podaffinity)
+    * [Story 3 - PodAntiAffinity](#story-3---podantiaffinity)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Design Details](#design-details)
+  * [Algorithm](#algorithm)
+    * [NodeAffinity](#nodeaffinity)
+    * [PodAffinity](#podaffinity)
+    * [PodAntiAffinity](#podantiaffinity)
+  * [Test Plan](#test-plan)
+  * [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Terms
+
+- **Topology:** describe a series of worker nodes which belongs to the same
+  region/zone/rack/hostname/etc. In terms of Kubernetes, they're defined and
+  grouped by node labels.
+- **Affinity**: if not specified particularly, "Affinity" refers to
+  `NodeAffinity`, `PodAffinity` and `PodAntiAffinity`.
+
+## Summary
+
+`EvenPodsSpreading` feature applies on NodeAffinity/PodAffinity/PodAntiAffinity
+to gives users more fine-grained control on distribution of pods scheduling, so
+as to achieve better high availability and resource utilization.
+
+## Motivation
+
+In Kubernetes, "Affinity" related directives are aimed to control how pods are
+scheduled - more packed or more scattering. But right now only limited options
+are offered: for `PodAffinity`, infinite pods can be stacked onto qualifying
+topology domain(s); for `PodAntiAffinity`, pods are scheduled onto each single
+topology domain exclusively.
+
+This is not an ideal situation if users want to put pods evenly across different
+topology domains - for the sake of high availability or saving cost. And regular
+rolling upgrade or scaling out replicas can also be problematic. See more
+details in [user stories](#user-stories).
+
+> Even pods spreading is a long-discussed topic, but it's only feasible to
+> hammer out the design/implementation details after recent Affinity performance
+> improvements.
+
+### Goals
+
+- Even spreading is achieved among pods, in the manner of NodeAffinity,
+  PodAffinity and PodAntiAffinity, and only impact
+  `RequiredDuringSchedulingIgnoredDuringExecution` affinity terms.
+- Even spreading is a predicate (hard requirement) instead of a priority (soft requirement).
+- Even spreading is implemented on limited topologies in initial version.
+
+### Non-Goals
+
+- Even spreading is NOT calculated on an application basis. In other words, it's
+  not _only_ applied within replicas of an application.
+- "Max number of pods per topology" is NOT a goal.
+- Scale-down on an application is not guaranteed to achieve even pods spreading
+  in initial implementation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - NodeAffinity
+
+As an application developer, I want my application pods to be scheduled onto
```

Comment:
Question:

What do you see as an expected behavior if node distribution is not uniform among zones.?
E.g one has 10 nodes from zone A and B and just 1 from zone C. (and lets assume we cannot put more nodes in C at the moment due to e.g. stockout)
And we have a spike on deployment with defined NodeAffinity over zone domain? 

I am asking because it is not obvious what should happen. We probably do not want to block scheduling in such case as it is better to have pods running somewhere (even suboptimally) than fail to serve users.

Also if we decide to use just nodes from A and B the situation will not improve over time. Pods will not be rescheduled to C if nodes from C later appear in the cluster.

Reply:
Good question. I'm afraid based on current framework, even this feature can't handle the situation _perfectly_.

A remedy is to set `maxSkew` to a reasonable value so that it's still allowed that pods are scheduled more onto A & B, but it's also possible the imbalance can happen between A & B.

A _possible_ solution to achieve "absolute fairness" is to take nodes number in each zone into consideration. For example, if the nodes # in each zone is 5/5/1, then a pod has 5/11 odds to land on A or B, and 1/11 odds to land on C. But IMO this is way too complicated, and the cost of "absolute fairness" outweighs its benefits.

Reply:
If we go with what I suggested above, one could use "preferred" spreading or use "MaxSkew" to address the imbalance issue. 

> A possible solution to achieve "absolute fairness" is to take nodes number in each zone into consideration. For example, if the nodes # in each zone is 5/5/1, then a pod has 5/11 odds to land on A or B, and 1/11 odds to land on C. But IMO this is way too complicated, and the cost of "absolute fairness" outweighs its benefits.

This is not necessary. If spreading based on the number of nodes was desired, one could use "node" as the topology key, instead of "zone". This is not exactly the same as what you suggested, but for practical scenarios it achieves similar results.

Reply:
I think making this a predicate(or hard requirement) will need developers/cluster-admins to have deeper knowledge of their app and say if we enable HPA, where new pods are scaled up/down, the `MaxSkew` need to be tweaked depending on the replicas etc, so as not having a scheduling failure.

Reply:
Scale down is good as no scheduling is involved.

WRT scale up, if we make it a predicate, new pod will be allowed to land on the nodes which satisfy formula "replicas num - minReplicas num < MaxSkew". For example, an unbalanced distribution as 3/3/0, and MaxSkew is 2, node3 is always a candidate, and node1/node2 can only be a candidate when it's distributed more balanced - e.g. 3/3/2. So generally this should be also fine.

Reply:
Do you think that if there unequal distribution of nodes in each zone/topology, people might still be ok with equal distribution of pods, even though in the last zone, all pods land on the same node?  Allowing that behavior via whatever configuration we expose will be helpful

Reply:
@krmayankk I think Bobby's https://github.com/kubernetes/enhancements/pull/851#discussion_r262714535 answers your question. In the case you described, it seems you're more concerned on balance of pods among nodes, rather than zones, right?

Reply:
@bsalamat I agree. For the first case, it's easier to be inferred that the 3rd domain is the only candidate (`0 - 0 < 1`, and `0` is the minMatch number and behaves globally).

While for the second case, it's vague, even experienced developers may need to stop a while to think about what the reasonable semantics/behavior is. Is the minMatch number `0`? If so, then we would get a scheduling failure. But if we think of the cluster as "4/3/~0~/~0~", then `3` is the minMatch num and the algorithm makes sense.

Reply:
@krmayankk yeap, a new section under "MaxSkew" section has detailed explanation now.